tests/unit/Core/Tool/VerifierTest.php

```diff
@@ -0,0 +1,119 @@
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = ['data' => 'things'];
+		$signature = $this->makeSig($this->sharedSecret, $url, json_encode($payload));
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->verified($signature, $this->apiKey, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertTrue($return);
+	}
+
+	public function testInvalidApiKeySignature()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = ['data' => 'things'];
+		$signature = $this->makeSig($this->sharedSecret, $url, json_encode($payload));
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(false);
+
+		$return = $verifier->verified($signature, $this->apiKey, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertFalse($return);
+	}
+
+	public function testInvalidSecretSignature()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = ['data' => 'things'];
+		$signature = $this->makeSig('notthesecret', $url, json_encode($payload));
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->verified($signature, $this->apiKey, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertFalse($return);
+	}
+
+	public function testInvalidDataSignature()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = ['data' => 'things'];
+		$signature = $this->makeSig($this->sharedSecret, $url, json_encode(['tampered' => 'data']));
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->verified($signature, $this->apiKey, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertFalse($return);
+	}
+
+	public function testCheckSignature()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+		$url = "http://localhost:8000/api/v3/exports/external/count/1?api_key={$this->apiKey}";
+		$payload = ['data' => 'things'];
+		$signature = $this->makeSig($this->sharedSecret, $url, json_encode($payload));
+
+		$return = $verifier->checkSignature($signature, $this->sharedSecret, $url, json_encode($payload));
+
+		$this->assertTrue($return);
+	}
+
+	public function testCheckApiKey()
+	{
+		$repo = M::mock(ApiKeyRepository::class);
+		$verifier = new Verifier($repo);
+
+		$repo->shouldReceive('apiKeyExists')->with($this->apiKey)->andReturn(true);
+
+		$return = $verifier->checkApiKey($this->apiKey);
+
+		$this->assertTrue($return);
+	}
+}
```

Comment:
It might be helpful to add test case(s) where the payload is an empty array, such as in a GET with long querystring, which is apparently also a valid situation, but caused me some grief once.

Reply:
Ok. Can do

Reply:
tiny comment in case it helps: $data was by default an array (when doing the compute signature) so that's why we got the weird empty array appended to our urls even in empty get requests.